sdk/core/System.TypeSpec/src/PipelineOptions.cs

```diff
@@ -0,0 +1,99 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.ComponentModel;
+using System.Diagnostics;
+using System.ServiceModel.Rest.Core;
+using System.Threading;
+
+namespace System.ServiceModel.Rest;
+
+// Make options freezable
+public class PipelineOptions
+{
+    private object _lock = new object();
+    private MessagePipeline? _pipeline;
+    private volatile bool _isFrozen;
+
+    private IPipelinePolicy<PipelineMessage>? _retryPolicy;
+    private IPipelinePolicy<PipelineMessage>? _loggingPolicy;
+    private PipelineTransport<PipelineMessage> _transport;
+    private ReadOnlyMemory<IPipelinePolicy<PipelineMessage>> _perTryPolicies;
+    private ReadOnlyMemory<IPipelinePolicy<PipelineMessage>> _perCallPolicies;
+
+    public PipelineOptions(PipelineTransport<PipelineMessage> transport)
+    {
+        _transport = transport;
+    }
+
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    public virtual MessagePipeline GetPipeline()
+    {
+        if (_pipeline != null) return _pipeline;
+
+        lock (_lock) {
```

Comment:
Why do we need to synchronize the read if we don't need to synchronize the write?

Reply:
Oh, I think I see now.  Is it because we're putting a lock around the pipeline creation operation?